src/inmanta/agent/handler.py

```diff
@@ -961,15 +1019,113 @@ def execute(self, ctx: HandlerContext, resource: TPurgeableResource, dry_run: bo
         super().execute(ctx, resource, dry_run)
 
 
+@stable_api
+class DiscoveryHandler(HandlerAPI, Generic[R, D]):
+    """
+    The DiscoveryHandler is generic with regard to two resource types:
+        - R denotes the handler's Discovery Resource type, used to drive resource discovery. This is not a
+          conventional resource type expected to be deployed on a network, but rather a way to express
+          the intent to discover resources of the second type D already present on the network.
+        - D denotes the handler's Unmanaged Resource type. This is the type of the resources that have been
+          discovered and reported to the server. Objects of this type must be serializable.
+    """
+
+    def deploy(
+        self,
+        ctx: HandlerContext,
+        resource: R,
+        requires: abc.Mapping[ResourceIdStr, ResourceState],
+    ) -> None:
+        self.execute(ctx, resource)
```

Comment:
We may want to copy over some of the this code

```
 resources_in_unexpected_state = filter_resources_in_unexpected_state(requires)
        if resources_in_unexpected_state:
            ctx.set_status(const.ResourceState.skipped)
            ctx.warning(
                "Resource %(resource)s skipped because a dependency is in an unexpected state: %(unexpected_states)s",
                resource=resource.id.resource_version_str(),
                unexpected_states=str({rid: state.value for rid, state in resources_in_unexpected_state.items()}),
            )
            return

        failed_dependencies = [req for req, status in requires.items() if status != ResourceState.deployed]
        if not any(failed_dependencies):
            self.execute(ctx, resource)
            if _should_reload():
                self.do_reload(ctx, resource)
        else:
            ctx.set_status(const.ResourceState.skipped)
            ctx.info(
                "Resource %(resource)s skipped due to failed dependencies: %(failed)s",
                resource=resource.id.resource_version_str(),
                failed=str(failed_dependencies),
            )
 ```
 
 To handle skip and failed dependencies

Reply:
I lifted this logic in the HandlerAPI's deploy in the other PR